File: android/src/main/kotlin/com/mytiki/sdk/capture/receipt/capacitor/js/JSStringType.kt
/*
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in the root directory.
 */

package com.mytiki.sdk.capture.receipt.capacitor.js

import com.getcapacitor.JSObject
import com.microblink.core.StringType
import org.json.JSONObject

/**
 * Represents a response object for a string type recognized by the TIKI SDK.
 *
 * @param stringType The [StringType] recognized by the TIKI SDK.
 */
class JSStringType(stringType: StringType)  {
    /**
     * The confidence score of the recognized string.
     */
    private val confidence: Float

    /**
     * The recognized string value.
     */
    private val value: String?

    init {
        confidence = stringType.confidence()
        value = stringType.value()
    }

    /**
     * Converts the [JSStringType] object to a JSON representation.
     *
     * @return A [JSONObject] containing the confidence and value of the recognized string.
     */
    fun toJS(): JSObject =
        JSObject()
            .put("confidence", confidence)
            .put("value", value)

    companion object {
        /**
         * Creates an [JSStringType] object from the provided [StringType].
         *
         * @param stringType The [StringType] to convert.
         * @return An [JSStringType] object representing the provided [StringType], or null if the input is null.
         */
        fun opt(stringType: StringType?): JSStringType? =
            if (stringType != null) JSStringType(stringType) else null
    }
}

File: ios/Plugin/Plugin/Js/JSStringType.swift
/*
 * RspStringType Struct
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

import Foundation
import BlinkReceipt
import BlinkEReceipt
import Capacitor

/**
 Represents a response containing a string value with confidence.

 This struct is used to convey a string value along with its confidence level.
 */
struct JSStringType{
    /// The confidence level of the string value.
    private let confidence: Float
    /// The string value.
    private let value: String?

    /**
     Initializes an `RspStringType` struct.

     - Parameters:
        - stringType: The string value with confidence information.
     */
    init(stringType: BRStringValue) {
        confidence = stringType.confidence
        value = stringType.value
    }

    /**
     Converts the `RspStringType` struct into a dictionary suitable for use in plugin response data.

     - Returns: A dictionary containing the string value and its confidence level in a format suitable for a Capacitor plugin call result.
     */
    public func toPluginCallResultData() -> Capacitor.PluginCallResultData {
        return [
            "confidence": confidence,
            "value": value ?? ""
        ]
    }
}

